### ISSUE TITLE
`make test` wrapper script

### ISSUE THREAD

**Issue by danielshahaf** [zsh-users/zsh-syntax-highlighting#262]
I usually use the following wrapper script to run `make test` and display failures only. Should it be added to the tree?

```sh
#!/bin/sh
# TODO: exit code 0 if PASS, 1 if FAIL/XPASS, 2 if harness died
make test \
| perl -E '@x = split /^(?=#)/m, `cat`; print for grep { /^ok.*# TODO/m or /^not ok(?!.*# TODO)/m } @x' \
| zsh -f tests/tap-colorizer.zsh
```

**Comment by phy1729** [comment]
Perhaps it'd be better to add this as a flag to `tap-colorizer.zsh`, so that say `make test failure_only=1` would only print failures. Diff below to show proof of concept.

    diff --git a/tests/tap-colorizer.zsh b/tests/tap-colorizer.zsh
    index cb645c6..c58fff9 100755
    --- a/tests/tap-colorizer.zsh
    +++ b/tests/tap-colorizer.zsh
    @@ -37,8 +37,9 @@ if [[ ! -t 1 ]] ; then
       exec cat
     fi
     
    -while read -r line;
    -do
    +local print_line
    +while read -r line; do
    +  print_line=$(( ! failure_only ))
       case $line in
         # comment (filename header) or plan
         (#* | <->..<->)
    @@ -46,6 +47,7 @@ do
           ;;
         # XPASS
         (ok*# TODO*)
    +      print_line=1
           print -nP %F{red}
           ;;
         # XFAIL
    @@ -54,6 +56,7 @@ do
           ;;
         # FAIL
         (not ok*)
    +      print_line=1
           print -nP %F{red}
           ;;
         # PASS
    @@ -61,7 +64,7 @@ do
           print -nP %F{green}
           ;;
       esac
    -  print -nr - "$line"
    +  (( print_line )) && print -nr - "$line"
       print -nP %f
    -  echo "" # newline
    +  (( print_line )) && echo "" # newline
     done

**Comment by danielshahaf** [comment]
> Perhaps it'd be better to add this as a flag to `tap-colorizer.zsh`, so that say `make test failure_only=1` would only print failures.

That's what I thought at first, too, but @phy1729 suggested it was a bad idea because `tap-colorizer.zsh` should do one thing and do it well: convert uncolored TAP output to colored TAP output.  (And I was convinced by his point.)

So perhaps this "hide anything but failures" should be implemented in a separate filter script?

**Comment by phy1729** [comment]
Did I suggest that? The problem with having two scripts is then we have two tap parsers instead of just one. Perhaps its one thing should expand to formatting tap instead of just coloring it.

**Comment by danielshahaf** [comment]
The parser is a single `case` statement with four branches, I'm not really concerned about duplicating that.  I think the win in code readability is worth the small duplication.

(Furthermore, the format it parses is specified by Perl, i.e., isn't likely to change.)

**Comment by danielshahaf** [comment]
... and if we want to avoid duplicating the parser, we could implement one or both filters in terms of [TAP::Parser](http://search.cpan.org/dist/TAP-Parser/lib/TAP/Parser.pm).

Here's a proof of concept for the "two filters" approach:

```diff
diff --git a/tests/tap-filter b/tests/tap-filter
new file mode 100755
index 0000000..72e8c56
--- /dev/null
+++ b/tests/tap-filter
@@ -0,0 +1,42 @@
+#!/usr/bin/env perl
+# -------------------------------------------------------------------------------------------------
+# Copyright (c) 2015 zsh-syntax-highlighting contributors
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without modification, are permitted
+# provided that the following conditions are met:
+#
+#  * Redistributions of source code must retain the above copyright notice, this list of conditions
+#    and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright notice, this list of
+#    conditions and the following disclaimer in the documentation and/or other materials provided
+#    with the distribution.
+#  * Neither the name of the zsh-syntax-highlighting contributors nor the names of its contributors
+#    may be used to endorse or promote products derived from this software without specific prior
+#    written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+# IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+# FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+# IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+# OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# -------------------------------------------------------------------------------------------------
+# -*- mode: zsh; sh-indentation: 2; indent-tabs-mode: nil; sh-basic-offset: 2; -*-
+# vim: ft=zsh sw=2 ts=2 et
+# -------------------------------------------------------------------------------------------------
+
+# This is a stdin-to-stdout filter that takes TAP output (such as 'make test')
+# on stdin and deletes lines not pertaining to unexpected failures.
+
+use v5.10.0;
+use warnings;
+use strict;
+
+undef $/; # slurp mode
+print for
+  grep { /^ok.*# TODO/m or /^not ok(?!.*# TODO)/m }
+    split /^(?=#)/m,
+      <STDIN>;
diff --git a/tests/test-highlighting.zsh b/tests/test-highlighting.zsh
index 0abf34b..aa1d5ca 100755
--- a/tests/test-highlighting.zsh
+++ b/tests/test-highlighting.zsh
@@ -130,6 +130,6 @@ run_test() {
 # Process each test data file in test data directory.
 integer something_failed=0
 for data_file in ${0:h:h}/highlighters/$1/test-data/*.zsh; do
-  run_test "$data_file" | tee >(${0:A:h}/tap-colorizer.zsh) | grep -v '^not ok.*# TODO' | grep -q '^not ok\|^ok.*# TODO' && (( something_failed=1 ))
+  run_test "$data_file" | tee >(${0:A:h}/tap-filter | ${0:A:h}/tap-colorizer.zsh) | grep -v '^not ok.*# TODO' | grep -q '^not ok\|^ok.*# TODO' && (( something_failed=1 ))
   (( $pipestatus[1] )) && exit 2
 done
```

**Comment by danielshahaf** [comment]
Fixed in  	9b64ad7 after discussion on IRC.
